0005-private-communities.md

```diff
@@ -0,0 +1,87 @@
+
+## Database
+
+The `CommunityVisibility` enum used in `community.visibility` gets a new variant `Private`. Queries in `post_view.rs` and `comment_view.rs` need to be adjusted so that content in communities set to private is only shown to followers. This also needs to be covered by unit tests.
+
+There needs to be a new table to store follow requests:
+```sql
+CREATE TABLE community_follow_request (
+    id serial PRIMARY KEY,
+    person_id int REFERENCES person ON UPDATE CASCADE ON DELETE CASCADE NOT NULL UNIQUE,
+    moderator_id int REFERENCES person ON UPDATE CASCADE ON DELETE CASCADE,
+    published timestamptz NOT NULL DEFAULT now()
+);
+```
+
+## API
+
+New follow requests for a private community need to be stored in `community_follow_request` table, and not in `community_follower`. This works very similar to the `registration_application` used for site registrations. Moderators can review and approve/reject applications with the following endpoints:
+- `GET /api/v3/community/follow_request/count`
+- `GET /api/v3/community/follow_request/list`
+- `POST /api/v3/community/follow_request/approve`
+
+The follow_request list for each item should contain a boolean value `is_new_instance`. This value should be true if the community has no followers from the user's instance yet, and allows showing a warning when content will be federated to a new server.
+
+Once the approve endpoint is called, the follow request can be deleted. If the request was approved, write the follow relation to `community_follower`. If it was rejected, we could notify the user about this (optional). 
+
+Users who don't follow the private community need to be prevented from posting or making other actions in it. This can be done by adding a check in [check_community_user_action()](https://github.com/LemmyNet/lemmy/blob/main/crates/api_common/src/utils.rs#L171).
+
+When [processing mentions](https://github.com/LemmyNet/lemmy/blob/main/crates/utils/src/utils/mention.rs#L24), the code needs to check if it is inside a private community, and in that case ignore mentions of users which don't follow it.
+
+The new endpoints should be covered by [Typescript API tests](https://github.com/LemmyNet/lemmy/tree/main/api_tests).
+
+## Federation
+
+As described in the API section, follow requests for private communities are stored in `community_follow_request` instead of `community_follower`. Once a request is approved, send an `Accept` activity back to the user.
+
+The `Group` actor for private communities remains public and can be fetched without restrictions, so that remote users can fetch it and follow it. It needs a new attribute to indicate that the group is not public. [ActivityStreams](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-group) doesn't seem to provide anything suitable, so we can simply add a custom attribute `private: true`.
```

Comment:
I couldn't find anything usable either.

Reply:
The micro-blogging part of the Fediverse uses `manuallyApprovesFollowers` property (boolean). It is not defined in AS vocabulary, and usually defined as `{"manuallyApprovesFollowers": "as:manuallyApprovesFollowers"}`. However, it is supported by most projects (including Mastodon).

Reply:
That makes sense, Ive changed it. Only thing is that it doesnt give any indication that content in the group is private.

Reply:
Perhaps `audience` property can be used for that. `as:Public` would indicate a public or semi-private community.

Feels a bit awkward because `audience` is already used in FEP-1b12 objects.